.github/workflows/build-extra-widgets.yml

```diff
@@ -0,0 +1,190 @@
+        run: |
+          BASE_TAG="${{ inputs.base_image_tag }}"
+          echo "Updating Dockerfile to use base image: ${BASE_TAG}"
+          sed -i "s|FROM .*|FROM ${BASE_TAG}|" Dockerfile
+          echo "Updated Dockerfile:"
+          cat Dockerfile
+
+      - name: Verify base image exists
+        run: |
+          echo "Verifying base image: ${{ inputs.base_image_tag }}"
+          docker manifest inspect ${{ inputs.base_image_tag }} || {
+            echo "ERROR: Base image does not exist or is not accessible"
+            exit 1
+          }
+      - name: Commit Dockerfile changes
+        run: |
+          git config user.name "github-actions[bot]"
+          git config user.email "github-actions[bot]@users.noreply.github.com"
+            
+          git add Dockerfile
+            
+          git diff --staged --quiet && echo "No changes" || {
+            git commit -m "Update base image to ${{ inputs.base_image_tag }}"
+            git push origin main
+            echo " Pushed Dockerfile changes"
+            }
+      - name: Log in to GitHub Container Registry
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Build and push image
+        run: |
+          # Create tag in format: branch-shortsha
+          TAG_NAME="${{ inputs.branch }}-${{ inputs.short_sha }}"
+          echo "Building image with tags: ${TAG_NAME} and latest"
+          
+          # Pull cache if available
+          docker pull $CI_REGISTRY_IMAGE:${TAG_NAME} || true
+          docker pull $CI_REGISTRY_IMAGE:latest || true
+          
+          # Build with cache
+          docker build \
+            --cache-from $CI_REGISTRY_IMAGE:${TAG_NAME} \
+            --cache-from $CI_REGISTRY_IMAGE:latest \
+            --tag $CI_REGISTRY_IMAGE:${TAG_NAME} \
+            --tag $CI_REGISTRY_IMAGE:latest \
+            .
+          
+          docker push $CI_REGISTRY_IMAGE:${TAG_NAME}
+          docker push $CI_REGISTRY_IMAGE:latest
+          
+          echo "Successfully pushed images:"
+          echo "   - $CI_REGISTRY_IMAGE:${TAG_NAME}"
+          echo "   - $CI_REGISTRY_IMAGE:latest"
+
+      - name: Store image info
+        id: image-info
+        run: |
+          TAG_NAME="${{ inputs.branch }}-${{ inputs.short_sha }}"
+          echo "tag=${TAG_NAME}" >> $GITHUB_OUTPUT
+
+      - name: Output image details
+        run: |
+          echo "Build complete!"
+          echo "Base image: ${{ inputs.base_image_tag }}"
+          echo "Enketo commit: ${{ inputs.short_sha }}"
+          echo "Branch: ${{ inputs.branch }}"
+          echo "Image tag: ${{ inputs.branch }}-${{ inputs.short_sha }}"
+
+  update-deployment-config:
+    needs: build
+    runs-on: ubuntu-latest
+    steps:
+      - name: Generate GitHub App Token
+        id: app-token
+        uses: actions/create-github-app-token@v1
```

Comment:
GitHub App Token & checkout & git settings are copy/pasted twice in this file, better extract as an action.

I think you can just copy/paste this [action](https://github.com/kobotoolbox/kpi/blob/18c96461bf8c988edc2c9076fd4513cf637b3318/.github/actions/checkout-with-github-app-token/action.yml) from KPI, see the [PR](https://github.com/kobotoolbox/kpi/pull/6374/files) how it's used ;) That will also solve 3 other comments along the way.